File: Data Structures & Algorithms/balanced-binary-tree/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
private:
    int dfs(TreeNode* node, int len, bool& isBalanced) {
        if (node == nullptr){
            return len;
        }       
        
        int left = dfs(node->left, len+1, isBalanced);
        int right = dfs(node->right, len+1, isBalanced);
        int diff = abs(left - right);
        isBalanced = isBalanced && (diff <= 1);
        return max(left, right);
    }

public:
    bool isBalanced(TreeNode* root) {
        bool isBalanced = true;
        dfs(root, 0, isBalanced);

        return isBalanced;
    }
};

File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        return self.serialize(subRoot) in self.serialize(root)


    def serialize(self, node: Optional[TreeNode]) -> str:
        if not node:
            return '#'
        return ("$" + str(node.val) + self.serialize(node.left) + self.serialize(node.right))